drivers/flash/nxp_qspi_nor.c

```diff
@@ -0,0 +1,419 @@
+
+	if (len % DT_PROP(SOC_NV_FLASH_NODE, erase_block_size)) {
+		sectors++;
+	}
+
+	for (int i = 0; i < sectors; i++) {
+		off_t offset_to_sector = offset + i * DT_PROP(SOC_NV_FLASH_NODE, erase_block_size);
+		status = flash_mcux_flexspi_qspi_write_enable(dev, offset);
+		if (status == kStatus_Success) {
+			flash_transfer.deviceAddress = offset_to_sector;
+			flash_transfer.port = kFLEXSPI_PortA1;
+			flash_transfer.cmdType = kFLEXSPI_Command;
+			flash_transfer.SeqNumber = 1;
+			flash_transfer.seqIndex = NOR_CMD_LUT_SEQ_IDX_ERASESECTOR;
+
+			status = FLEXSPI_TransferBlocking(base_address, &flash_transfer);
+			if (status == kStatus_Success) {
+				status = flash_mcux_flexspi_qspi_wait_bus_busy(dev, offset);
+			}
+		}
+	}
+
+	irq_unlock(key);
+	k_sem_give(&priv->write_lock);
+
+	return status;
+}
+
+/* Write/erase operations in this driver is protected by a semaphore. This
+ * prevents access from multiple threads, but using this function the semaphore
+ * can be locked, preventing all write/erase operations. */
+static int flash_mcux_flexspi_qspi_write_protection(struct device *dev, bool enable)
+{
+	struct flash_priv *priv = dev->driver_data;
+	int rc = 0;
+
+	if (enable) {
+		rc = k_sem_take(&priv->write_lock, K_FOREVER);
+	} else {
+		k_sem_give(&priv->write_lock);
+	}
+
+	return rc;
+}
+
+#if defined(CONFIG_FLASH_PAGE_LAYOUT)
+static const struct flash_pages_layout dev_layout = {
+	.pages_count = KB(CONFIG_FLASH_SIZE) /
+		       DT_PROP(SOC_NV_FLASH_NODE, erase_block_size),
+	.pages_size = DT_PROP(SOC_NV_FLASH_NODE, erase_block_size),
+};
+
+static void flash_mcux_flexspi_qspi_pages_layout(struct device *dev,
+						 const struct flash_pages_layout **layout,
+						 size_t *layout_size)
+{
+	*layout = &dev_layout;
+	*layout_size = 1;
+}
+#endif /* CONFIG_FLASH_PAGE_LAYOUT */
+
+static struct flash_priv flash_data;
+
+static const struct flash_driver_api flash_mcux_flexspi_qspi_api = {
+	.write_protection = flash_mcux_flexspi_qspi_write_protection,
+	.erase = flash_mcux_flexspi_qspi_erase,
+	.write = flash_mcux_flexspi_qspi_write,
+	.read = flash_mcux_flexspi_qspi_read,
+#if defined(CONFIG_FLASH_PAGE_LAYOUT)
+	.page_layout = flash_mcux_flexspi_qspi_pages_layout,
+#endif
+	.write_block_size = DT_PROP(SOC_NV_FLASH_NODE, write_block_size),
+};
+
+static int flash_mcux_flexspi_qspi_init(struct device *dev)
+{
+	// Get the parent node of the chosen flash chip to get the reg address of the FlexSPI device
+	FLEXSPI_Type *base = (FLEXSPI_Type *)DT_REG_ADDR_BY_IDX(DT_INST(0, nxp_imx_flexspi), 0);
+	struct flash_priv *priv = dev->driver_data;
+
+	priv->pflash_block_base = base;
+	unsigned int key;
+
+	key = irq_lock();
+
+	/* Switch to PLL2 for XIP to avoid hardfault during re-initialize clock. */
+	CLOCK_InitSysPfd(kCLOCK_Pfd2, 24);      /* Set PLL2 PFD2 clock 396MHZ. */
+	CLOCK_SetMux(kCLOCK_FlexspiMux, 0x2);   /* Choose PLL2 PFD2 clock as flexspi source clock. */
+	CLOCK_SetDiv(kCLOCK_FlexspiDiv, 2);     /* flexspi clock 133M. */
+
```

Comment:
it's better to move clock configuration to soc.c, since it might diff from soc to soc.

Reply:
Yes! Will look into that. Thanks